File: src/p1456_maximum_number_of_vowels_in_a_substring_of_given_length.cpp
#include <algorithm>
#include <catch2/catch_test_macros.hpp>
#include <string>

namespace P1456 {
    constexpr inline bool is_vowel(char c) {
        return (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u');
    }
    class Solution {
      public:
        int maxVowels(const std::string &s, const int &k) {
            int num_vowels =
                std::count_if(s.cbegin(), s.cbegin() + k, is_vowel);
            int max_vowels = num_vowels;

            for (std::size_t i = 0, j = k; j < s.size(); ++i, ++j) {
                if (is_vowel(s[i])) {
                    --num_vowels;
                }
                if (is_vowel(s[j])) {
                    ++num_vowels;
                }
                max_vowels = std::max(max_vowels, num_vowels);
            }
            return max_vowels;
        }
    };

    TEST_CASE("1456 - Maximum Number of Vowels in a Substring of Given Length",
              "[String][Sliding Window]") {

        SECTION("Example 1") {
            // Input: s = "abciiidef", k = 3
            // Output: 3
            Solution s;
            REQUIRE(s.maxVowels("abciiidef", 3) == 3);
        }

        SECTION("Example 2") {
            // Input: s = "aeiou", k = 2
            // Output: 2
            Solution s;
            REQUIRE(s.maxVowels("aeiou", 2) == 2);
        }

        SECTION("Example 3") {
            // Input: s = "leetcode", k = 3
            // Output: 2
            Solution s;
            REQUIRE(s.maxVowels("leetcode", 3) == 2);
        }
    }
} // namespace P1456

File: src/p1456_maximum_number_of_vowels_in_a_substring_of_given_length.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowels = {"a", "e", "i", "o", "u"}

        num_vowels = sum(1 for c in s[:k] if c in vowels)
        max_vowels = num_vowels

        for i in range(len(s) - k):
            num_vowels -= s[i] in vowels
            num_vowels += s[i + k] in vowels
            max_vowels = max(max_vowels, num_vowels)

        return max_vowels


def test_1456():
    s = Solution()

    # Example 1:
    # Input: s = "abciiidef", k = 3
    # Output: 3
    assert s.maxVowels("abciiidef", 3) == 3

    # Example 2:
    # Input: s = "aeiou", k = 2
    # Output: 2
    assert s.maxVowels("aeiou", 2) == 2

    # Example 3:
    # Input: s = "leetcode", k = 3
    # Output: 2
    assert s.maxVowels("leetcode", 3) == 2
